sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,324 @@
+            infer_batch_size: batch size for workers.
+            show_progress_bar: whether to display a progress bar
+
+        Returns:
+            samples: posterior samples obtained by performing inference on xs
+                given the posterior
+
+        """
+        num_sim_samples = xs.shape[0]
+        xs_batches = torch.split(xs, infer_batch_size, dim=0)
+
+        if num_workers != 1:
+            # Parallelize the sequence of batches across workers.
+            # We use the solution proposed here: https://stackoverflow.com/a/61689175
+            # to update the pbar only after the workers finished a task.
+            with tqdm_joblib(
+                tqdm(
+                    xs_batches,
+                    disable=not show_progress_bar,
+                    desc=f"Performing {num_sim_samples} posterior runs in"
+                    f"{len(xs_batches)} batches.",
+                    total=len(xs_batches),
+                )
+            ) as _:
+                samples: Tensor
+                samples = Parallel(n_jobs=num_workers)(  # pyright: ignore[reportAssignmentType]
+                    delayed(infer_posterior_on_batch)(
+                        xs_batch, posterior, num_posterior_samples
+                    )
+                    for xs_batch in xs_batches
+                )
+        else:
+            pbar = tqdm(
+                total=num_sim_samples,
+                disable=not show_progress_bar,
+                desc=f"Running {num_sim_samples} samples for tarp analysis.",
+            )
+
+            with pbar:
+                samples = []
+                for xs_batch in xs_batches:
+                    samples.append(
+                        infer_posterior_on_batch(
+                            xs_batch, posterior, num_posterior_samples
+                        )
+                    )
+                    pbar.update(infer_batch_size)
+                samples = torch.cat(samples, dim=1)
+
+        return samples
+
+    def check(
+        self,
+        samples: Tensor,
+        theta: Tensor,
+    ) -> Tuple[Tensor, Tensor]:
+        """
+        Estimates coverage with the TARP method.
+        Reference: `Lemos, Coogan et al 2023 <https://arxiv.org/abs/2302.03026>`_
+
+        Args:
+            samples: The predicted parameter samples to compute the coverage of,
+                     with shape ``(n_samples, n_sims, n_dims)``. These are
+                     obtained by sampling a trained posterior.  Multiple
+                     (posterior) samples for one observation are encouraged.
+            theta: The true parameter value theta, with shape ``(n_sims, n_dims)``.
+
+        Returns:
+            ecp: Expected coverage probability (``ecp``)
+            alpha: credibility values
+
+        """
+        # TARP assumes that the predicted thetas are sampled from the "true"
+        # PDF num_samples times
+        theta = theta.detach() if len(theta.shape) != 2 else theta.detach().unsqueeze(0)
```

Comment:
can't we assert already here that `theta.shape == samples.shape[1:]`? Why do we need the `unsqueeze(0)`?

Reply:
I wanted to handle the data shapes early on. I guess asserting the shapes is one way of going.